### PR TITLE
listener: extract active_tcp_socket and active_stream_listener_base

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:

Split from #16947 
Extract ActiveTcpSocket from ActiveTcpListener.
Extract a base class ActiveStreamListener from ActiveTcpListener.

Additional Description:
The new base class owns the the active sockets that drive itself through the listener filters. 
After the active socket passes all the listener filters, a server connection is created. As a derived listener,
ActiveTcpListener overrides `newActiveConnection` to take the ownership of that server connection.

Risk Level: LOW
Testing: 
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by mattklein123** [comment]
Check CI?

/wait

**Comment by lambdai** [comment]
Ahrr, Add a setting to my vscode

**Review by mattklein123 — COMMENTED** [review]
Can you summarize what code you actually changed vs. moved? What do I need to review?

/wait-any

**Comment by lambdai** [comment]
> Can you summarize what code you actually changed vs. moved? What do I need to review?
> 
> /wait-any

Sure. 
1. active_tcp_socket.{h,cc} It's a pure move other than referring a parent class of ActiveTcpListener.
2. ActiveStreamListenerBase is the new extracted parent class of ActiveTcpListener. Sitting in it's new file active_stream_listener_base.{h, cc}. Its interface a strict subset of  the previous ActiveTcpListener. The only exception is `newActiveConnection` that is designed to be overridden by ActiveTcpListener.
Particularly, no data member is added to ActiveTcpListener+ActiveStreamListenerBase

File: source/server/active_stream_listener_base.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+   * @param filter_chain The network filter chain linking to the connection.
+   * @param server_conn_ptr The server connection.
+   * @param stream_info The stream info of the active connection.
+   */
+  virtual void newActiveConnection(const Network::FilterChain& filter_chain,
+                                   Network::ServerConnectionPtr server_conn_ptr,
+                                   std::unique_ptr<StreamInfo::StreamInfo> stream_info) PURE;
+  Event::Dispatcher& dispatcher_;
+  Network::ListenerPtr listener_;
+  bool is_deleting_{false};
```
Comment (mattklein123) [inline source/server/active_stream_listener_base.h:111]: Related to my other comment it's a little odd that this is protected and used by derived classes. Can you add comments here on the expectations?

File: source/server/active_stream_listener_base.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        }
+      }
+    }
+  }
+
+  // Below members are open to access by ActiveTcpSocket.
+  Network::ConnectionHandler& parent_;
+  const std::chrono::milliseconds listener_filters_timeout_;
+  const bool continue_on_listener_filters_timeout_;
+  std::list<std::unique_ptr<ActiveTcpSocket>> sockets_;
```
Comment (mattklein123) [inline source/server/active_stream_listener_base.h:95]: The 3 before this are effectively  const, so I'm OK with being public, but this one should probably not be wide open? I would have accessors around this?

File: source/server/active_stream_listener_base.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+   * active connection.
+   *
+   * @param filter_chain The network filter chain linking to the connection.
+   * @param server_conn_ptr The server connection.
+   * @param stream_info The stream info of the active connection.
+   */
+  virtual void newActiveConnection(const Network::FilterChain& filter_chain,
+                                   Network::ServerConnectionPtr server_conn_ptr,
+                                   std::unique_ptr<StreamInfo::StreamInfo> stream_info) PURE;
+  Event::Dispatcher& dispatcher_;
```
Comment (mattklein123) [inline source/server/active_stream_listener_base.h:109]: Technically this can be private or just lose the accessor above.

File: source/server/active_tcp_listener.h [hunk 4]
```diff
@@ -71,48 +75,45 @@ class ActiveTcpListener final : public Network::TcpListenerCallbacks,
                       bool hand_off_restored_destination_connections, bool rebalanced) override;
 
   /**
-   * Remove and destroy an active connection.
-   * @param connection supplies the connection to remove.
+   * Create active connection from the server connection.
```
Comment (mattklein123) [inline source/server/active_tcp_listener.h:78]: nit, don't need doc comments for overriden function.

File: source/server/active_tcp_listener.h [hunk 5]
```diff
@@ -71,48 +75,45 @@ class ActiveTcpListener final : public Network::TcpListenerCallbacks,
...
+   * Update the listener config. The follow up connections will see the new config. The existing
+   * connections are not impacted.
    */
-  ActiveConnections& getOrCreateActiveConnections(const Network::FilterChain& filter_chain);
+  void updateListenerConfig(Network::ListenerConfig& config);
 
   /**
-   * Schedule to remove and destroy the active connections which are not tracked by listener
-   * config. Caution: The connection are not destroyed yet when function returns.
+   * Schedule to remove and destroy the active connection owned by the filter chain.
```
Comment (mattklein123) [inline source/server/active_tcp_listener.h:96]:
```suggestion
   * Schedule removal and destruction of all active connections owned by a filter chain.
```

?

File: source/server/active_tcp_listener.h [hunk 6]
```diff
@@ -71,48 +75,45 @@ class ActiveTcpListener final : public Network::TcpListenerCallbacks,
...
    */
-  void
-  deferredRemoveFilterChains(const std::list<const Network::FilterChain*>& draining_filter_chains);
+  void deferRemoveFilterChain(const Network::FilterChain* filter_chain) override;
```
Comment (mattklein123) [inline source/server/active_tcp_listener.h:98]: I would probably just call this `removeFilterChain()` as I think the deferral part doesn't really matter in this context.

File: source/server/active_tcp_listener.h [hunk 7]
```diff
@@ -71,48 +75,45 @@ class ActiveTcpListener final : public Network::TcpListenerCallbacks,
...
-  void updateListenerConfig(Network::ListenerConfig& config);
+  void removeConnection(ActiveTcpConnection& connection);
 
-  Network::TcpConnectionHandler& parent_;
-  Network::ListenerPtr listener_;
-  const std::chrono::milliseconds listener_filters_timeout_;
-  const bool continue_on_listener_filters_timeout_;
-  std::list<ActiveTcpSocketPtr> sockets_;
-  absl::flat_hash_map<const Network::FilterChain*, ActiveConnectionsPtr> connections_by_context_;
+  absl::node_hash_map<const Network::FilterChain*, std::unique_ptr<ActiveConnections>>
```
Comment (mattklein123) [inline source/server/active_tcp_listener.h:106]: Why was this changed from flat to node hash map?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, generally makes sense with some random comments.

/wait

**Comment by lambdai** [comment]
fixing conflict
/wait

File: source/server/active_stream_listener_base.h (hunk 2 above)
Comment (lambdai) [reply source/server/active_stream_listener_base.h:95]: It is public because half of the class ActiveTcpListener is extracted to this base. Moved to protected. Is it good enough?

Comment (mattklein123) [reply source/server/active_stream_listener_base.h:95]: Sure moving to protected seems better, thanks.

/wait

File: source/server/active_tcp_listener.h (hunk 7 above)
Comment (mattklein123) [reply source/server/active_tcp_listener.h:101]: Hmm, this should work, what is the compile error you get?

Comment (lambdai) [reply source/server/active_tcp_listener.h:101]: OK... I don't see error in this PR right now. 
Revert to `flat_hash_map`. If I see it in the next PR we can discuss with concrete error.

**Comment by mattklein123** [comment]
OK LGTM pending the other open comment thread about moving to protected, thanks.

/wait

File: source/server/active_tcp_listener.h (hunk 7 above)
Comment (lambdai) [reply source/server/active_tcp_listener.h]: I had the illusion it is done.

Now I am trying to move to protected and I realize it was public even before this PR.

This things I am doing is:
1. declare it protected.
2. add const getter. it is for test only, I can take a look how to avoid using it in the test.
3. add a remove method in this class to mutate this list.

**Comment by lambdai** [comment]
> OK LGTM pending the other open comment thread about moving to protected, thanks.
> 
> /wait

I had the illusion it is done.

Now I am trying to move to protected and I realize it was public even before this PR.

This things I am doing is:
1. declare it protected.
2. add const getter. it is for test only, I can take a look how to avoid using it in the test.
3. add a remove method in this class to mutate this list.

**Review by mattklein123 — APPROVED** [review]
Thanks!
